### PR TITLE
Retiled matrix + refactoring bt_band_to_tridiag

### PR DESCRIPTION
**Pull request by rasolca**
Note:
  Relevant performance improvement of the backtransformation when applied to only few eigenvectors.
  For the current use cases the difference is irrelevant.

### CODE REVIEW

File: include/dlaf/eigensolver/bt_band_to_tridiag/impl.h [hunk 1]
```diff
@@ -160,85 +165,109 @@ struct ApplyHHToSingleTileRow<Backend::GPU, T> {
...
-    const SizeType fullsize = 2 * b - 1;
-    const SizeType mb = dist.blockSize().rows();
-    const SizeType rows_below = dist.size().rows() - offset.row();
+    rows_v_ = std::min(2 * b, dist_e.size().rows() - offset.row()) - 1;
+    rows_v_top_ = std::min(rows_v_, b - 1);
 
     // Note:
-    // In general, keep in mind that to sub_offset.row() is applied a + 1, to take into account the
-    // offset about how reflectors are actually applied.
+    // In general, keep in mind that the first row of the itop tile_e should not be touched.
```
Comment (msimberg) [inline include/dlaf/eigensolver/bt_band_to_tridiag/impl.h:269]:
```suggestion
    // In general, keep in mind that the first row of the top tile_e should not be touched.
```
?

File: include/dlaf/eigensolver/bt_band_to_tridiag/impl.h [hunk 2]
```diff
@@ -849,84 +852,83 @@ void BackTransformationT2B<B, D, T>::call(comm::CommunicatorGrid grid, const Siz
...
+            auto subtile_e_ro = splitTile(mat_e_rt.read(idx_e), spec_e);
+
+            // W2 = V* . E
+            ex::start_detached(
+                dlaf::internal::whenAllLift(blas::Op::ConjTrans, blas::Op::NoTrans, T(1),
+                                            keepFuture(subtile_v), keepFuture(subtile_e_ro), T(0),
+                                            splitTile(mat_w2tmp(idx_w2), helper.specW2(nb))) |
+                dlaf::tile::gemm(dlaf::internal::Policy<B>(thread_priority::normal)));
+
+            subtile_e_ro = {};  // Clear the shared future to free dependencies.
```
Comment (msimberg) [inline include/dlaf/eigensolver/bt_band_to_tridiag/impl.h:914]:
```suggestion
                                            keepFuture(subtile_v), keepFuture(std::move(subtile_e_ro)), T(0),
                                            splitTile(mat_w2tmp(idx_w2), helper.specW2(nb))) |
                dlaf::tile::gemm(dlaf::internal::Policy<B>(thread_priority::normal)));
```
?

File: include/dlaf/matrix/matrix_base.h [hunk 3]
```diff
@@ -33,10 +49,15 @@ class MatrixBase {
   }
 
   /// Returns the block size of the matrix.
-  const TileElementSize& blockSize() const noexcept {
+  const TileElementSize blockSize() const noexcept {
```
Comment (msimberg) [inline include/dlaf/matrix/matrix_base.h:52]: Is the reference gone intentionally? If yes, why not:
```suggestion
  TileElementSize blockSize() const noexcept {
```
?

File: include/dlaf/matrix/retiled_matrix.h [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+#include "dlaf/matrix/tile.h"
+#include "dlaf/sender/keep_future.h"
+#include "dlaf/types.h"
+
+#include "dlaf/common/range2d.h"
+#include "dlaf/common/vector.h"
+
+namespace dlaf::matrix {
+
+/// TODO
```
Comment (msimberg) [inline include/dlaf/matrix/retiled_matrix.h:30]: Still to do?

File: include/dlaf/matrix/tile.h [hunk 5]
```diff
@@ -226,6 +239,12 @@ class Tile<const T, D> {
     return data_.ld() == data_.size().rows();
   }
 
+  /// Returns a subtile.
+  /// Note: to avoid segfaults or race conditions, the original tile must be kept in scope.
+  Tile subTileReference(const SubTileSpec& spec) const noexcept {
+    return Tile(*this, spec);
+  }
+
```
Comment (msimberg) [inline include/dlaf/matrix/tile.h:247]: I'm just trying to check my understanding of this:

This is _almost_ a copy constructor for `Tile` in the sense that the underlying memory chunk is correctly reference counted and kept alive, but the original tile is what keeps subsequent accesses from happening, and _that_ part is not reference counted?

**Review by msimberg — APPROVED** [review]
I can't comment on the algorithmic changes. Only a few superficial comments, but this looks good to me.

File: include/dlaf/matrix/tile.h (hunk 5 above)
Comment (rasolca) [reply include/dlaf/matrix/tile.h:247]: Summary of the separate discussion: This method is used to allow the usage of tile methods in tasks that execute more than one operation and do not work on full tiles.
The fact that the underlying memory is reference counted and kept alive is not important for this method (it is only an implementation detail and should not be relied on).

**Review by albestro — CHANGES_REQUESTED** [review]
Nice work!

About switching to `tile` (+ `tile_per_block`) and not directly deal anymore with `block` might be more extensive (e.g. `Matrix` still uses `block_size` as initialisation parameter).

In case we agree on that, we might just take that into account for the (near) future refactoring.

File: include/dlaf/matrix/retiled_matrix.h [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+public:
+  static constexpr Device device = D;
+
+  using ElementType = T;
+  using TileType = Tile<ElementType, D>;
+  using ConstTileType = Tile<const ElementType, D>;
+  using TileDataType = internal::TileData<const ElementType, D>;
+
+  ///
+  /// @pre matrix.blockSize() is divisible by tiles_per_block.
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/retiled_matrix.h:50]: Don't think so... We have many examples in the code without such a comment.

File: test/include/dlaf_test/matrix/util_matrix.h [hunk 7]
```diff
@@ -92,7 +92,7 @@ namespace internal {
 /// @pre The second argument of err_message should be either T, T& or const T&.
 template <template <class, Device> class MatrixType, class T, class ElementGetter, class ComparisonOp,
           class ErrorMessageGetter>
-void check(ElementGetter expected, MatrixType<const T, Device::CPU>& mat, ComparisonOp comp,
+void check(ElementGetter expected, MatrixType<T, Device::CPU>& mat, ComparisonOp comp,
```
[the remark this answers is not in the document]
Comment (albestro) [reply test/include/dlaf_test/matrix/util_matrix.h:95]: As discussed offline, my point was about const-correctness. We will take this to the refactoring discussion 😉

File: include/dlaf/matrix/distribution.h [hunk 8]
```diff
@@ -84,8 +100,13 @@ class Distribution {
...
   }
 
-  const TileElementSize& blockSize() const noexcept {
-    return block_size_;
+  const TileElementSize blockSize() const noexcept {
+    return TileElementSize{tile_size_.rows() * tiles_per_block_.rows(),
+                           tile_size_.cols() * tiles_per_block_.cols()};
+  }
+
+  const TileElementSize& baseTileSize() const noexcept {
```
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/matrix/distribution.h:108]: Yep, understood.

- `blockSize()` -> distribution block size
- `baseTileSize()` -> full size of a tile in general (old block size)
- `tileSize(idx)` -> similar to baseTileSize, but it actually returns the size of a specific tile (e.g. incomplete ones)

File: include/dlaf/matrix/distribution.h [hunk 9]
```diff
@@ -39,14 +39,30 @@ class Distribution {
   /// @param[in] rank_index is the rank of the current process,
   /// @param[in] source_rank_index is the rank of the process which contains the top left tile of the matrix,
   /// @pre size.isValid(),
-  /// @pre !block_size.isEmpty(),
+  /// @pre !tile_size.isEmpty(),
```
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/matrix/distribution.h:42]: Ok, it was a wrong deduction from the previous status of this PR. After the quick discussion we had, I agree with your point.

**Review by albestro — APPROVED** [review]
Just a couple of typos. LGTM otherwise 👍🏻

File: include/dlaf/matrix/matrix_base.h [hunk 10]
```diff
@@ -22,7 +22,22 @@ namespace internal {
...
+  }
+
+  MatrixBase(const Distribution& distribution, const LocalTileSize& tiles_per_block)
+      : distribution_(std::make_shared<Distribution>(  //
+            distribution.size(), distribution.blockSize(),
+            TileElementSize{distribution.blockSize().rows() / tiles_per_block.rows(),
+                            distribution.blockSize().cols() / tiles_per_block.cols()},
+            distribution.commGridSize(), distribution.rankIndex(), distribution.sourceRankIndex())) {
+    DLAF_ASSERT(distribution.blockSize() == distribution.baseTileSize(),
+                "distribution should be the distribution of the original Matrix.");
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/matrix_base.h:37]: Yep, currently I do not support retiling a retiled matrix.

File: include/dlaf/matrix/matrix_base.h [hunk 11]
```diff
@@ -22,7 +22,22 @@ namespace internal {
 class MatrixBase {
 public:
   MatrixBase(Distribution distribution)
-      : distribution_(std::make_shared<Distribution>(std::move(distribution))) {}
+      : distribution_(std::make_shared<Distribution>(std::move(distribution))) {
+    DLAF_ASSERT(distribution.blockSize() == distribution.baseTileSize(),
+                "Multi Tile distribution block is not supperted by Matrix yet.");
```
[the remark this answers is not in the document]
Comment (rasolca) [reply include/dlaf/matrix/matrix_base.h:27]: MatrixBase supports it through the second constructor (https://github.com/eth-cscs/DLA-Future/blob/5c5acee7c3007f150abd1db4feb5df7916fd8c15/include/dlaf/matrix/matrix_base.h#L30).
This constructor is what is used by Matrix which do not support retiled distribution.

This choice was made not to over-engineer things that will be heavily refactored due to the matrix objects unification.

**Comment by rasolca** [comment]
cscs-ci run

**Pull request merged**